File: 修订补充示例/第二轮/ch10_正文代码框.R
# 正文代码框顺序运行；扩展数据凡带“模拟”均非真实调查记录。
args <- commandArgs(trailingOnly=FALSE)
file_arg <- args[grepl("^--file=",args)]
script_dir <- if(length(file_arg)) dirname(normalizePath(gsub("~+~"," ",sub("^--file=","",file_arg[1]),fixed=TRUE))) else getwd()
repo_dir <- normalizePath(file.path(script_dir,"..",".."))
local_lib <- file.path(repo_dir,"..",".R-library")
if(dir.exists(local_lib)) .libPaths(c(local_lib,.libPaths()))
setwd(repo_dir)
library(sandwich)
library(lmtest)
set.seed(922)
df <- read.csv(file.path(repo_dir,"Chapter10_Prediction_ch_prediction/results/chapter10_nbs_70city_prediction_data.csv"),fileEncoding="UTF-8-BOM")

# 代码框 1: {R 中训练/测试预测}
data_path <- file.path("Chapter10_Prediction_ch_prediction",
                       "results", "chapter10_nbs_70city_prediction_data.csv")
df <- read.csv(data_path, fileEncoding="UTF-8-BOM")
train <- subset(df, month <= 9)
test  <- subset(df, month >= 10)
mod <- lm(second_hand_yoy ~ new_house_yoy + second_hand_mom + new_house_mom + first_tier,
          data = train)
pred <- predict(mod, newdata = test)
mean((test$second_hand_yoy - pred)^2)


# 代码框 2: {R 中按组诊断测试误差}
test$error <- test$second_hand_yoy - pred
test$sq_error <- test$error^2
aggregate(sq_error ~ month, data = test, mean)
aggregate(sq_error ~ first_tier, data = test, mean)


# 代码框 3: {R 中均值基准}
test$pred_train_mean <- mean(train$second_hand_yoy, na.rm = TRUE)
mean((test$second_hand_yoy - test$pred_train_mean)^2)

city_mean <- aggregate(second_hand_yoy ~ city, data = train, mean)
names(city_mean)[2] <- "pred_city_mean"
test_with_mean <- merge(test, city_mean, by = "city", all.x = TRUE)
mean((test_with_mean$second_hand_yoy - test_with_mean$pred_city_mean)^2)


# 代码框 4: {R 中随机森林}
library(ranger)
train_df <- train[, c("second_hand_yoy", "new_house_yoy", "second_hand_mom",
                       "new_house_mom", "first_tier")]
rf <- ranger(second_hand_yoy ~ ., data = train_df, num.trees = 500,
             mtry = 3, importance = "permutation")
# OOB MSE（ranger 默认口径，越小越好）
rf$prediction.error
# 置换重要性
importance(rf)

File: 修订补充示例/第二轮/ch10_正文代码框.py
# 正文代码框顺序运行；带“模拟”的扩展数据不是真实调查记录。
from pathlib import Path
import numpy as np,pandas as pd,statsmodels.formula.api as smf
from scipy import stats
repo_dir=Path(__file__).resolve().parents[2]
import os
os.chdir(repo_dir)
df=pd.read_csv(repo_dir / 'Chapter10_Prediction_ch_prediction/results/chapter10_nbs_70city_prediction_data.csv')

# 代码框 1: {Python 中训练/测试预测}
import numpy as np, pandas as pd
import statsmodels.formula.api as smf
from pathlib import Path
data_path = (Path("Chapter10_Prediction_ch_prediction")
             / "results" / "chapter10_nbs_70city_prediction_data.csv")
df = pd.read_csv(data_path)
train = df[df["month"] <= 9]
test = df[df["month"] >= 10]
model = smf.ols(
    "second_hand_yoy ~ new_house_yoy + second_hand_mom + new_house_mom + first_tier",
    data=train
).fit()
pred = model.predict(test)
((test["second_hand_yoy"] - pred) ** 2).mean()


# 代码框 2: {Python 中避免标准化泄漏}
from sklearn.preprocessing import StandardScaler
xvars = ["new_house_yoy", "second_hand_mom", "new_house_mom", "first_tier"]
scaler = StandardScaler().fit(train[xvars])
X_train = scaler.transform(train[xvars])
X_test = scaler.transform(test[xvars])
# 不要对 train 和 test 合并后再 fit scaler


# 代码框 3: {Python 中简单基准模型}
test["pred_train_mean"] = train["second_hand_yoy"].mean()
mse_mean = ((test["second_hand_yoy"] - test["pred_train_mean"]) ** 2).mean()

city_mean = train.groupby("city")["second_hand_yoy"].mean()
test["pred_city_mean"] = test["city"].map(city_mean)
mse_city = ((test["second_hand_yoy"] - test["pred_city_mean"]) ** 2).mean()


# 代码框 4: {Python 中随机森林}
from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor(n_estimators=500, max_features="sqrt",
                           oob_score=True, random_state=42)
y_train = train["second_hand_yoy"].to_numpy()
rf.fit(X_train, y_train)
rf.oob_score_  # OOB R²（sklearn 默认口径，越大越好）
rf.feature_importances_  # 回归树的加权平方误差下降重要性

